velox/docs/develop/spilling.rst

```diff
@@ -0,0 +1,480 @@
+4. Spill input rows to disk if the corresponding partitions have been spilled by the build side which are identified by child spill partition id set if set;
+5. Join the non-spilled probe input rows with the hash table and produce the output result;
+6. After all the operators finish processing the inputs, if there is no spilled data needed to restore, then all the hash probe operators finish. Otherwise, the last finished operator will send the next spill input to the hash build operators through the hash join bridge;
+7. Go back to step1 to repeat the join process with the next built hash table;
+
+Note that we need to disable some hash probe optimizations if the spilling has
+been triggered by the hash build. For example, we can’t apply dynamic filtering
+optimization in that case as we don’t have the complete set of join keys from a
+partially built hash table.
+
+We also don't support spilling for `null-aware anti-join type <https://github.com/facebookincubator/velox/blob/main/velox/core/PlanNode.h#L931>`_ with filter set as
+it requires to cross join null-key probe rows with all build-side rows for
+filter evaluation to check if the null-key probe rows can be added to output
+or not.
+
+`HashJoinBridge <https://github.com/facebookincubator/velox/blob/main/velox/exec/HashJoinBridge.h#L28>`_
+^^^^^^^^^^^^^^
+The HashJoinBridge object also needs the following extensions to support the
+spilling:
+
+* extends the existing `setHashTable <https://github.com/facebookincubator/velox/blob/main/velox/exec/HashJoinBridge.h#L41>`_ interface to take optional spilled
+  partition metadata if spilling has been triggered while building the table;
+* adds `probeFinished <https://github.com/facebookincubator/velox/blob/main/velox/exec/HashJoinBridge.h#L83>`_ interface for the hash probe operator to set and notify
+  the hash build operators of the spill input to build the next hash table;
+* adds `spillInputOrFuture <https://github.com/facebookincubator/velox/blob/main/velox/exec/HashJoinBridge.h#L102>`_ interface for the hash build operator to wait for the
+  spill input to build the next hash table;
+* Internally, the object needs to maintain all the spill partitions remaining
+  to restore in an ordered map and restore the next spill partition from the
+  beginning of the map. To ensure the child partitions are restored first, we
+  add the SpillPartitionId type as the key in the map to identify a spill
+  partition which consists of partition bit offset and the partition number.
+  The partition with higher bit offset is placed ahead of the one with lower
+  bit offset. If the partition bit offset ties, then partition with a lower
+  partition number goes first;
+* To parallelize the hash table build from the spilled partition, the hash join
+  bridge will split the spill partition files among the hash build operators
+  with each one having an equally-sized shard to restore.
```

Comment:
nit: Perhaps, add a Future Work section to mention memory arbitration, performance evaluation, etc.